File: AndroidClient/app/src/main/java/com/mancersoft/litevpn/transport/UdpTransport.java
package com.mancersoft.litevpn.transport;

import android.util.Log;

import com.mancersoft.litevpn.VpnManager;

import java.io.Closeable;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ClosedByInterruptException;
import java.nio.channels.ClosedChannelException;
import java.nio.channels.DatagramChannel;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import static com.mancersoft.litevpn.VpnManager.TAG;

public class UdpTransport implements IVpnTransport {

    private DatagramChannel mUdpChannel;

    private final ExecutorService mExecutorService = Executors.newCachedThreadPool();

    private final String mServerName;
    private final int mServerPort;

    private Future<?> mIncomingMessageProcessing;

    private MessageListener mMessageListener;
    private ClosedListener mClosedListener;

    public UdpTransport(String serverName, int serverPort) {
        mServerName = serverName;
        mServerPort = serverPort;
    }

    @Override
    public Closeable createSocket() throws IOException {
        mUdpChannel = DatagramChannel.open();
        return mUdpChannel.socket();
    }

    @Override
    public boolean isReliable() {
        return false;
    }

    @Override
    public void setOnMessageListener(MessageListener messageListener) {
        mMessageListener = messageListener;
    }

    @Override
    public void setOnClosedListener(ClosedListener closedListener) {
        mClosedListener = closedListener;
    }

    @Override
    public CompletableFuture<Boolean> connect() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                disconnect();
                SocketAddress serverAddress = new InetSocketAddress(mServerName, mServerPort);
                mUdpChannel.connect(serverAddress);
                mUdpChannel.configureBlocking(true);
                mIncomingMessageProcessing = mExecutorService.submit(this::incomingMessagesProcessing);
                return true;
            } catch (IOException e) {
                Log.e(TAG, "UdpTransport connect error", e);
                return false;
            }
        });
    }

    private void disconnect(boolean isByUser) {
        disconnect();
        if (mClosedListener != null) {
            mClosedListener.onClosed(isByUser);
        }
    }

    @Override
    public void disconnect() {
        try {
            if (mIncomingMessageProcessing != null) {
                mIncomingMessageProcessing.cancel(true);
            }

            if (mUdpChannel != null && mUdpChannel.isConnected()) {
                mUdpChannel.close();
            }
        } catch (Exception e) {
            Log.e(TAG, "UdpTransport disconnect error", e);
        }
    }

    @Override
    public void sendAsync(Packet packetToSend) {
        Packet packet = new Packet(packetToSend);
        mExecutorService.execute(() -> {
            try {
                send(packet.getData(), packet.getLength());
            } catch (ClosedChannelException e1) {
                disconnect(true);
            } catch (Exception e2) {
                Log.e(TAG, "UdpTransport sendAsync error", e2);
                disconnect(false);
            }
        });
    }

    private void send(byte[] data, int length) throws IOException {
        ByteBuffer buffer = ByteBuffer.wrap(data, 0, length);
        mUdpChannel.write(buffer);
    }

    private void receive(Packet packet) throws IOException {
        ByteBuffer buffer = ByteBuffer.wrap(packet.getData());
        packet.setLength(mUdpChannel.read(buffer));
    }

    private void incomingMessagesProcessing() {
        try {
            Packet packet = new Packet();
            packet.setData(new byte[VpnManager.MAX_PACKET_SIZE]);
            while (!Thread.currentThread().isInterrupted()) {
                //System.out.println("Start receive transport");
                this.receive(packet);
                //System.out.println("Stop receive transport");
                if (mMessageListener != null) {
                    mMessageListener.onMessage(new Packet(packet));
                }
            }
        } catch (ClosedByInterruptException e1) {
            disconnect(true);
        } catch (Exception e2) {
            Log.e(TAG, "UdpTransport incomingMessagesProcessing error", e2);
            disconnect(false);
        }
    }
}

File: AndroidClient/app/src/main/kotlin/com/mancersoft/litevpn/transport/UdpTransport.kt
package com.mancersoft.litevpn.transport

import android.util.Log
import com.mancersoft.litevpn.MAX_PACKET_SIZE
import com.mancersoft.litevpn.TAG
import java.io.Closeable
import java.io.IOException
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.nio.ByteBuffer
import java.nio.channels.ClosedByInterruptException
import java.nio.channels.ClosedChannelException
import java.nio.channels.DatagramChannel
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.Future

class UdpTransport(private val mServerName: String, private val mServerPort: Int) : IVpnTransport {

    private var mUdpChannel: DatagramChannel? = null
    private val mExecutorService = Executors.newCachedThreadPool()
    private var mIncomingMessageProcessing: Future<*>? = null
    private var mMessageListener: ((Packet) -> Unit)? = null
    private var mClosedListener: ((isByUser: Boolean) -> Unit)? = null

    @Throws(IOException::class)
    override fun createSocket(): Closeable {
        disconnect()
        mUdpChannel = DatagramChannel.open()
        return mUdpChannel!!.socket()
    }

    override val isReliable: Boolean
        get() = false

    override fun setOnMessageListener(messageListener: ((Packet) -> Unit)?) {
        mMessageListener = messageListener
    }

    override fun setOnClosedListener(closedListener: ((isByUser: Boolean) -> Unit)?) {
        mClosedListener = closedListener
    }

    override fun connect(): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync {
            try {
                disconnect()
                val serverAddress: SocketAddress = InetSocketAddress(mServerName, mServerPort)
                mUdpChannel!!.connect(serverAddress)
                mUdpChannel!!.configureBlocking(true)
                mIncomingMessageProcessing = mExecutorService.submit { incomingMessagesProcessing() }
                return@supplyAsync true
            } catch (e: IOException) {
                Log.e(TAG, "UdpTransport connect error", e)
                return@supplyAsync false
            }
        }
    }

    private fun disconnect(isByUser: Boolean) {
        disconnect()
        mClosedListener?.invoke(isByUser)
    }

    override fun disconnect() {
        try {
            mIncomingMessageProcessing?.cancel(true)
            if (mUdpChannel != null && mUdpChannel!!.isConnected) {
                mUdpChannel!!.close()
            }
        } catch (e: Exception) {
            Log.e(TAG, "UdpTransport disconnect error", e)
        }
    }

    override fun sendAsync(packet: Packet) {
        val packetToSend = Packet(packet)
        mExecutorService.execute {
            try {
                send(packetToSend.data, packetToSend.length)
            } catch (e1: ClosedChannelException) {
                disconnect(true)
            } catch (e2: Exception) {
                Log.e(TAG, "UdpTransport sendAsync error", e2)
                disconnect(false)
            }
        }
    }

    @Throws(IOException::class)
    private fun send(data: ByteArray, length: Int) {
        val buffer = ByteBuffer.wrap(data, 0, length)
        mUdpChannel!!.write(buffer)
    }

    @Throws(IOException::class)
    private fun receive(packet: Packet) {
        val buffer = ByteBuffer.wrap(packet.data)
        packet.length = mUdpChannel!!.read(buffer)
    }

    private fun incomingMessagesProcessing() {
        try {
            val packet = Packet()
            packet.data = ByteArray(MAX_PACKET_SIZE)
            while (!Thread.currentThread().isInterrupted) {
                //System.out.println("Start receive transport");
                receive(packet)
                //System.out.println("Stop receive transport");
                mMessageListener?.invoke(Packet(packet))
            }
        } catch (e1: ClosedByInterruptException) {
            disconnect(true)
        } catch (e2: Exception) {
            Log.e(TAG, "UdpTransport incomingMessagesProcessing error", e2)
            disconnect(false)
        }
    }

}